> Notebook-friendly copy of `part-I/1.2-data-structures-and-control-flow-exercises.ipynb`, generated by `tools/make_live.py`. Edit the book notebook, not this file.

In [1]:
# --- environment setup (generated, not part of the lesson) ---
# Colab and Kaggle do not ship every package this notebook imports.
# This is a no-op in an environment that is already set up.
import importlib.util
import subprocess
import sys

for module, package in {"pooch": "pooch"}.items():
    if importlib.util.find_spec(module) is None:
        subprocess.run([sys.executable, "-m", "pip", "install", "-q", package], check=True)

# Exercises

**ℹ️ How to use these exercises**

Attempt each from scratch in the empty code cell. Write plain functions with clear names, and avoid a mutable default argument.

## Exercise 1: Lists — build, modify, slice

Start from the daily mean temperatures `[-2.3, -1.1, 0.4, 1.2, -0.8]` (°C).

1. Append a sixth reading, `-3.1`.
2. Remove the value `0.4` (the sensor flagged it as unreliable).
3. Print the first reading, the last reading, and a slice containing the second and third.
4. Print how many readings remain.

In [2]:
daily_temperatures = [-2.3, -1.1, 0.4, 1.2, -0.8]
daily_temperatures.append(-3.1)
daily_temperatures.remove(0.4)
print(daily_temperatures[0], daily_temperatures[4])
print(daily_temperatures[1:3])
print(len(daily_temperatures))

-2.3 -3.1
[-1.1, 1.2]
5


## Exercise 2: Dictionaries — build, look up, remove

You are given station records as a list of tuples `(code, name, elevation_m)`:

```python
[("JFJ", "Jungfraujoch", 3571), ("BAS", "Basel-Binningen", 316), ("LUG", "Lugano", 273)]
```

1. Build a dict mapping each `code` to its `elevation_m`.
2. Print the elevation for `"JFJ"`.
3. Look up `"ZRH"` with `.get`, returning `"unknown"` if it is absent.
4. Delete the entry for `"LUG"`, then print each remaining code and its elevation using
   `.items()`.

In [3]:
stations = {
  "JFJ": { "name":"Jungfraujoch", "elevation_m": 3571 },
  "BAS" : { "name":"Basel-Binningen", "elevation_m": 316 },
  "LUG" : { "name":"Lugano", "elevation_m": 273 }
}
print(stations["JFJ"]["elevation_m"])
print(stations.get("ZRH ",{}).get("name", "unknown"))
del stations["LUG"]
print(stations.items())

3571
unknown
dict_items([('JFJ', {'name': 'Jungfraujoch', 'elevation_m': 3571}), ('BAS', {'name': 'Basel-Binningen', 'elevation_m': 316})])


## Exercise 3: A pure function

Write a function `mean(values)` (no type hints) that returns the average of a list of numbers. Test it on `[6.1, 7.4, 5.9, 8.2, 6.8]`, printing the result rounded to two decimals.

In [4]:
numbers= [6.1, 7.4, 5.9, 8.2, 6.8]
def mean(values):
  "return the mean of a list of values"
  total = 0.0
  for v in values:
    total+= v
  return total/len(values)
print(round(mean(numbers),2))

6.88


## Exercise 4: A function with a default argument

Write a function `is_warm(temp_celsius, threshold=10.0)` (no type hints) that returns `True` when the temperature is at or above the threshold and `False` otherwise, using `if`/`else`. Test it on 6.1 °C with the default threshold, and again on 6.1 °C with a threshold of 5.0 °C passed by keyword.

In [5]:
def is_warm(temp_celsius, threshold=10.0):
  if temp_celsius >= threshold :
     return True
  else:
    return False
print(is_warm(6.1))
print(is_warm(6.1, threshold = 5.0 ))

False
True


## Exercise 5: Pair two stations with zip

Two stations recorded the same five days:

```python
dates = ["Mon", "Tue", "Wed", "Thu", "Fri"]
jfj_celsius = [-2.3, -1.1, 0.4, 1.2, -0.8]
bas_celsius = [6.1, 7.4, 5.9, 8.2, 6.8]
```

Using `zip`, print one line per day showing the date and the temperature difference
(Basel minus Jungfraujoch), rounded to two decimals.

In [6]:
dates = ["Mon", "Tue", "Wed", "Thu", "Fri"]
jfj_celsius = [-2.3, -1.1, 0.4, 1.2, -0.8]
bas_celsius = [6.1, 7.4, 5.9, 8.2, 6.8]

for date, jfj, bas in zip(dates, jfj_celsius, bas_celsius):
  difference= bas - jfj
  print(f"{date}: {difference: .2f}°C")

Mon:  8.40°C
Tue:  8.50°C
Wed:  5.50°C
Thu:  7.00°C
Fri:  7.60°C


## Exercise 6: Number the readings with enumerate

A station wrote one reading per day for a week:

```python
week_celsius = [6.1, 7.4, 5.9, 8.2, 11.3, 9.8, 6.8]
```

1. Using `enumerate`, print one line per reading: the day number and the temperature. Count days
   from 1, not 0 — `enumerate` starts at 0, so the day number is the index plus one.
2. In the same loop, keep track of the warmest reading seen so far and the day it fell on, in two
   variables defined before the loop. Do not use `max`.
3. After the loop, print the warmest reading and its day number.

In [7]:
week_celsius = [6.1, 7.4, 5.9, 8.2, 11.3, 9.8, 6.8]
warmest_temp = week_celsius[0]
warmest_day = 1
for i, temp  in enumerate(week_celsius):
  day = i+1
  print(day,temp)
  if temp > warmest_temp:
    warmest_temp = temp
    warmest_day = day
print(f"Warmest: {warmest_temp} °C on day {warmest_day}")


1 6.1
2 7.4
3 5.9
4 8.2
5 11.3
6 9.8
7 6.8
Warmest: 11.3 °C on day 5


## Exercise 7: while, break, and continue

A logger returned six readings, two of which are missing:

```python
mixed_celsius = [6.1, None, 5.9, None, 9.5, 6.8]
```

Using a `while` loop over the indices:

1. Skip any missing value with `continue`.
2. Stop at the first reading above 9.0 °C with `break`, storing it in `first_hot`.
3. Print `first_hot` after the loop. Define it before the loop so the code is safe even if no
   reading exceeds the threshold.

In [13]:
mixed_celsius = [6.1, None, 5.9, None, 9.5, 6.8]
i = 0
first_hot = None
while i<len(mixed_celsius):
  temp = mixed_celsius[i]
  if temp is None:
    i+=1
    continue
  if temp > 9.0:
    first_hot = temp
    break
  i+=1
print(first_hot)

9.5


## Exercise 8: Default arguments and returning two values

Write a function `summarise(values, scale=1.0)` that multiplies every value by `scale` and
returns **two** things: the total and the mean.

1. Call it on `[6.1, 7.4, 5.9, 8.2, 6.8]` with the default scale, unpacking the result into
   `total_celsius` and `mean_celsius`.
2. Call it again with `scale=2.0` passed by keyword, and print only the mean.
3. Call it once more and store the result in a single name. Print its `type` — what did you get?

In [11]:
values = [6.1, 7.4, 5.9, 8.2, 6.8]
def summarise(values, scale = 1.0):
  total = 0
  for v in values:
    total+= v * scale
  return total, total/len(values)
  print(total/len(values))

total_celsius, mean_celsius = summarise(values)

total_scaled, mean_scaled = summarise(values, scale=2.0)
print(mean_scaled)

result = summarise(values)
print(type(result))

13.76
<class 'tuple'>


## Exercise 9: A variable number of arguments

Write a function `warmest(*temps_celsius)` (no type hints) that returns the highest value among
any number of positional arguments, using a loop and a running maximum (not `max`). Test it on
three individual readings passed directly, and again by unpacking the list
`[6.1, 7.4, 5.9, 8.2, 6.8]` with `*`.

In [10]:
def warmest(*temps_celsius):
    warmest_temp = temps_celsius[0]

    for temp in temps_celsius:
        if temp > warmest_temp:
            warmest_temp = temp

    return warmest_temp
print(warmest(6.1, 7.4, 5.9))
values = [6.1, 7.4, 5.9, 8.2, 6.8]

print(warmest(*values))

7.4
8.2


## Exercise 10: Fix the shared-default bug

An assistant wrote this function; it is correct the first time but wrong on later calls:

```python
def collect_above(values, threshold, out=[]):
    for v in values:
        if v > threshold:
            out.append(v)
    return out
```

In one comment, explain the bug, then rewrite it so each call starts fresh. Demonstrate with two
separate calls, assigning each result to its own variable.

In [9]:
# The default list is shared between calls if we use out=[], so using None lets us create a fresh list for each call.
def collect_above(values, threshold, out=None):
    if out is None:
        out = []

    for v in values:
        if v > threshold:
            out.append(v)

    return out
result_1 = collect_above([6.1, 7.4, 5.9, 8.2], 7.0)
result_2 = collect_above([3.2, 9.1, 4.5, 10.2], 8.0)

print(result_1)
print(result_2)

[7.4, 8.2]
[9.1, 10.2]


## Going deeper (optional)

### Exercise 11: Document your function with a docstring

Take the `mean(values)` you wrote in Exercise 3 and give it a docstring — a one-line summary in
triple quotes on the first line of the body. State what it returns *and* the unit it assumes, so
the function carries information the name alone cannot.

Then print `mean.__doc__`, and call `help(mean)` to see what a reader of your code would see.

In [15]:
numbers= [6.1, 7.4, 5.9, 8.2, 6.8]
def mean(values):
  """Return the mean of a list of temperatures in degrees Celsius."""

  total = 0.0
  for v in values:
    total+= v
  return total/len(values)
print(mean.__doc__)
help(mean)

Return the mean of a list of temperatures in degrees Celsius.
Help on function mean in module __main__:

mean(values)
    Return the mean of a list of temperatures in degrees Celsius.



### Exercise 12: Tuples — unpacking and immutability

A station's position is stored as a tuple, `station_coords = (46.5475, 7.9853)`.

1. Unpack it into `lat` and `lon` in one line, and print both.
2. Try to change the latitude to `47.0`. Catch the resulting error with `try`/`except TypeError`
   and print the message.
3. In a comment, say why a coordinate pair is better stored as a tuple than as a list.

In [18]:
station_coords = (46.5475, 7.9853)
lat,lon = station_coords
print(lat)
print(lon)
try:
    station_coords[0] = 47.0
except TypeError as error:
    print(error)
# A tuple is better for coordinates because the values should not be changed.

46.5475
7.9853
'tuple' object does not support item assignment


### Exercise 13: Means and a filter, with comprehensions

Recreate a `stations` dict where each code maps to a dict containing `readings_celsius`:

```python
{"JFJ": {"readings_celsius": [-2.3, -1.1, 0.4, 1.2, -0.8]},
 "BAS": {"readings_celsius": [6.1, 7.4, 5.9, 8.2, 6.8]},
 "LUG": {"readings_celsius": [9.4, 10.2, 8.8, 11.1, 9.9]}}
```

With a dict comprehension, build `means = {code: mean_temp}`, calling the `mean` function you
wrote in Exercise 3 rather than summing inline. Then, with a list comprehension, build the list
of codes whose mean is below 0 °C.

In [20]:
stations= {"JFJ": {"readings_celsius": [-2.3, -1.1, 0.4, 1.2, -0.8]},
 "BAS": {"readings_celsius": [6.1, 7.4, 5.9, 8.2, 6.8]},
 "LUG": {"readings_celsius": [9.4, 10.2, 8.8, 11.1, 9.9]}}

means = {
    code: mean(station["readings_celsius"])
    for code, station in stations.items()
}

def mean(values):
  """Return the mean of a list of temperatures in degrees Celsius."""

cold_codes = [
    code
    for code, mean_temp in means.items()
    if mean_temp < 0
]

print(means)
print(cold_codes)

{'JFJ': -0.52, 'BAS': 6.88, 'LUG': 9.879999999999999}
['JFJ']


### Exercise 14: Spot the impure function

This function is meant to return the readings without the last one:

```python
def drop_last(readings):
    readings.pop()
    return readings
```

1. Call it on `original = [6.1, 7.4, 5.9]`, then print `original`. What happened?
2. Rewrite it as a pure function that leaves the caller's list untouched, and demonstrate that
   `original` is unchanged after the call.
3. In one comment, connect this to the mutable-default bug from the lecture.

In [25]:
def drop_last(readings):
    new_readings = readings.copy()
    new_readings.pop()
    return new_readings

original = [6.1, 7.4, 5.9]

result = drop_last(original)

print(result)
print(original)
# Like the mutable-default bug, changing a shared list inside a function can affect data outside the function.

[6.1, 7.4]
[6.1, 7.4, 5.9]


### Exercise 15: Reject impossible temperatures

Write `to_kelvin(temp_celsius)` that raises a `ValueError` when the input is below absolute zero
(−273.15 °C) and otherwise returns the temperature in kelvin. Give it a docstring and a type hint.

Loop over `[20.0, -300.0, 0.0]`, calling it inside a `try`/`except` so that the invalid value is
reported without stopping the loop.

In [26]:
def to_kelvin(temp_celsius: float) -> float:
    """Convert Celsius to Kelvin and reject temperatures below absolute zero."""
    if temp_celsius < -273.15:
        raise ValueError("Temperature cannot be below absolute zero.")
    return temp_celsius + 273.15


for temp in [20.0, -300.0, 0.0]:
    try:
        kelvin = to_kelvin(temp)
        print(f"{temp} °C = {kelvin} K")
    except ValueError as error:
        print(f"Invalid temperature: {temp} °C — {error}")

20.0 °C = 293.15 K
Invalid temperature: -300.0 °C — Temperature cannot be below absolute zero.
0.0 °C = 273.15 K


## Exercise 16: A year of station data, continued

The optional section ends with Exercise 15. Exercises 16 and 17 are the main exercises of this
subchapter.

In subchapter 1.1 you fetched this file, worked out its structure, and converted a single
record to °C. You did not yet have the tools to handle all of it at once — now you do.

The goal: turn a year of daily maximum temperatures into a monthly summary, and write that
summary out as a new file. The input file has the same three columns as in Exercise 9 of the
previous subchapter: the station code, the day as `dd.mm.yy`, and the daily maximum temperature
in degrees Fahrenheit. Your monthly summary will have its own columns: `month`, `n_obs`
(the number of observations), and `mean_temp_celsius`.

In [27]:
# Pre-supplied: download the data file and cache it locally.
# You do not need to understand this cell yet — fetching data is covered in the
# reproducible-data-pipelines bonus subchapter.
import pooch

data_file = pooch.retrieve(
    url="https://raw.githubusercontent.com/gse-unil/2026_MLEES_book/main/data/part-I/station_iib_daily_max_temp_2022.csv",
    known_hash="sha256:034755fb289b4e157e5d029995481786a359eb25a80df62c09ee83d063013144",
    fname="station_iib_daily_max_temp_2022.csv",
    path=pooch.os_cache("mlees"),
)
print("data cached at:", data_file)

data cached at: /root/.cache/mlees/station_iib_daily_max_temp_2022.csv


**Step 1.** Write a function `fahrenheit_to_celsius(temp_fahrenheit)` that returns the
temperature in °C. Test it on the freezing point of water (32 °F should give 0.0 °C) and on
212 °F. A one-line docstring is optional here (see Exercise 11).

You wrote this conversion inline in 1.1. Steps 2 and 4 both need it again; with a function,
they call the same tested code instead of repeating the arithmetic.

In [28]:
def fahrenheit_to_celsius(temp_fahrenheit):
    """Convert a temperature from Fahrenheit to Celsius."""
    return (temp_fahrenheit - 32) * 5 / 9

print(fahrenheit_to_celsius(32))
print(fahrenheit_to_celsius(212))

0.0
100.0


**Step 2.** Open the file, read all lines, and loop over the records (skipping the header).
For each one, extract the month and the temperature in °C. Print the first three as a check.

Sixteen days in this file have an empty temperature field — a real gap in the station's record,
late October to mid-November. `float("")` raises `ValueError`, so your loop has to decide what to
do with those days before it can get to the end of the file. Skip them, and count how many you
skipped. After the loop, print how many records you converted and how many you skipped; the two
should add up to 365.

In [37]:
from pathlib import Path

skipped = 0
converted = 0

with Path(data_file).open("r", encoding="utf-8") as fhandler:
    lines = fhandler.readlines()

for line in lines[1:]:
    fields = line.strip().split(",")

    month = fields[1].split(".")[1]

    if fields[2] == "":
        skipped += 1
        continue

    temp_fahrenheit = float(fields[2])
    temp_celsius = fahrenheit_to_celsius(temp_fahrenheit)

    converted += 1

    if converted <= 3:
        print(month, temp_celsius)

print("converted:", converted)
print("skipped:", skipped)
print("total:", converted + skipped)

01 -9.61111111111111
01 -13.0
01 -4.222222222222223
converted: 349
skipped: 16
total: 365


**Step 3.** Build a dictionary mapping each month to the number of observations in it.

You do not know in advance which months are present, so the dictionary has to grow as you go.
Keep skipping the days with no reading: a day the sensor did not record is not an observation, so
October and November should come out short of their full length.

In [31]:
counts = {}

with Path(data_file).open("r", encoding="utf-8") as fhandler:
    lines = fhandler.readlines()

for line in lines[1:]:
    fields = line.strip().split(",")

    if fields[2] == "":
        continue

    month = fields[1].split(".")[1]

    counts[month] = counts.get(month, 0) + 1

print(counts)

{'01': 31, '02': 28, '03': 31, '04': 30, '05': 31, '06': 30, '07': 31, '08': 31, '09': 30, '10': 24, '11': 21, '12': 31}


**Step 4.** Build a second dictionary mapping each month to its **mean** daily maximum
temperature in °C. Print it with each mean rounded to two decimals.

In [32]:
counts = {}
totals = {}

with Path(data_file).open("r", encoding="utf-8") as fhandler:
    lines = fhandler.readlines()

for line in lines[1:]:
    fields = line.strip().split(",")

    if fields[2] == "":
        continue

    month = fields[1].split(".")[1]

    temp_fahrenheit = float(fields[2])
    temp_celsius = fahrenheit_to_celsius(temp_fahrenheit)

    counts[month] = counts.get(month, 0) + 1
    totals[month] = totals.get(month, 0) + temp_celsius

means = {}

for month in totals:
    means[month] = totals[month] / counts[month]

print(means)

{'01': -4.3584229390681015, '02': -0.2202380952380953, '03': 8.112903225806452, '04': 11.549999999999999, '05': 22.21505376344086, '06': 27.44814814814815, '07': 27.974910394265226, '08': 26.68100358422939, '09': 23.755555555555556, '10': 17.75694444444444, '11': 4.3994708994708995, '12': -1.3924731182795702}


**Step 5.** Find the warmest month and print it with its mean temperature.

Do this with a loop and a comparison, keeping track of the best value seen so far, as you did for
the warmest day in Exercise 6. A dictionary is looked up by key, not by position, so there is no
`[0]` to start from: start from `None` instead.

In [33]:
warmest_month = None
warmest_mean = None

for month, mean_temp in means.items():
    if warmest_mean is None:
        warmest_month = month
        warmest_mean = mean_temp
    elif mean_temp > warmest_mean:
        warmest_month = month
        warmest_mean = mean_temp

print("warmest month:", warmest_month)
print("mean temperature:", round(warmest_mean, 2))

warmest month: 07
mean temperature: 27.97


**Step 6.** Write your monthly summary to `_files/monthly_means.csv`, with the header
`month,n_obs,mean_temp_celsius` and one line per month, creating the `_files` folder first as you
did in 1.1. Then read the file back and print it, to confirm it says what you meant it to say.

In [35]:
summary_path = Path("_files/monthly_means.csv")
summary_path.parent.mkdir(exist_ok=True)

with summary_path.open("w", encoding="utf-8") as fhandler:
    fhandler.write("month,n_obs,mean_temp_celsius\n")

    for month in means:
        fhandler.write(
            f"{month},{counts[month]},{means[month]:.2f}\n"
        )

with summary_path.open("r", encoding="utf-8") as fhandler:
    summary = fhandler.read()

print(summary)

month,n_obs,mean_temp_celsius
01,31,-4.36
02,28,-0.22
03,31,8.11
04,30,11.55
05,31,22.22
06,30,27.45
07,31,27.97
08,31,26.68
09,30,23.76
10,24,17.76
11,21,4.40
12,31,-1.39



**Step 7 (optional).** This step uses `try`/`except` from the going-deeper box in the lecture. Real files contain junk, and you have already met one kind of it — the empty
temperature field. Suppose some lines are truncated as well, or hold a temperature that is not a
number at all. Modify your reading loop so that it skips any line that does not have three fields
or whose temperature will not convert to a number, counting how many lines it skipped.

Test it on this list before applying it to the real file:

```python
test_lines = [
    "IIB,01.01.22,30.2",
    "IIB,02.01.22",
    "IIB,03.01.22,notanumber",
    "IIB,04.01.22,33.4",
]
```

You should keep 2 and skip 2.

In [39]:
test_lines = [
    "IIB,01.01.22,30.2",
    "IIB,02.01.22",
    "IIB,03.01.22,notanumber",
    "IIB,04.01.22,33.4",
]
kept = 0
skipped = 0

for line in test_lines:
    fields = line.strip().split(",")

    if len(fields) != 3:
        skipped += 1
        continue

    try:
        temp_fahrenheit = float(fields[2])
    except ValueError:
        skipped += 1
        continue

    temp_celsius = fahrenheit_to_celsius(temp_fahrenheit)
    kept += 1

print("kept:", kept)
print("skipped:", skipped)

kept: 2
skipped: 2


**ℹ️ What you just did**

You wrote a reusable function, looped over a whole year of records, built dictionaries whose
keys you did not know in advance, aggregated by group, and wrote a derived dataset back to disk
— with a loop that survives malformed input. This is the shape of nearly every data-processing
script you will write. In subchapter 1.5, `pandas` will do the grouping in one line; the
reasoning will be exactly the same.

## Exercise 17: The solar system

<img src="https://raw.githubusercontent.com/gse-unil/2026_MLEES_book/main/part-I/_static/nasa-solar-system-montage.jpg" alt="The eight planets and the Moon photographed by spacecraft, arranged diagonally against black" width="100%">

<em>The eight planets and the Moon, as photographed by Mariner 10, Magellan, Galileo, Mars Global
Surveyor, Cassini and the Voyager spacecraft. The inner bodies are roughly to scale with each other, as are the outer ones, but the
two groups are not to scale with one another. Image
<a href="https://www.jpl.nasa.gov/images/pia03153-solar-system-montage-high-resolution-2001-version/">NASA/JPL (PIA03153)</a>, public domain.</em>


This is the long exercise for this subchapter, and the only one in it whose data is not
environmental. Eight planets and their masses are few enough to type by hand, and familiar enough
that a wrong answer is easy to spot, which makes them a good first test of lists, dictionaries and
functions together.

Work through sections A to C in order: section C writes functions against the dictionary you
build in section B. You will pick the exercise up again in 1.7, once classes are available, and
turn the dictionary into a `Planet` type.

The masses come from NASA's
[planetary fact sheet](https://nssdc.gsfc.nasa.gov/planetary/factsheet/). Use units of
10^24 kg throughout, so that Earth is 5.97 and Jupiter is 1898 — the variable names below carry
that unit, which is what makes a bare number like `1898` readable three cells later.

### A: lists and loops

**Q1) Create a list with the names of every planet in the solar system, in order.**

Name it `planets` and use lowercase names, so it starts `["mercury", "venus", ...]`.

In [40]:
planets= ["mercury", "venus", "earth", "mars", "jupiter", "saturn", "uranus", "neptune"]

**Q2) Have Python tell you how many planets there are by examining your list.**

You should get 8. Do not type the number — read it off the list.

In [42]:
len(planets)

8

**Q3) Use slicing to display the first four planets — the rocky ones.**

In [44]:
print(planets[0:4])

['mercury', 'venus', 'earth', 'mars']


**Q4) Iterate through your planets and print the planet name only if it ends in "s".**

You should see `venus`, `mars` and `uranus`, and nothing else.

Hint: the last letter of a name is `planet[-1]`. Use `continue` for the names that do not match,
so the loop body says explicitly what it skips.

In [49]:
for planet in planets:
  if planet[-1]!="s":
    continue
  print(planet)

venus
mars
uranus


### B: dictionary

**Q5) Create a dictionary that maps each planet name to its mass.**

Call it `planet_masses_1e24kg`, and read the values off the
[NASA fact sheet](https://nssdc.gsfc.nasa.gov/planetary/factsheet/) in units of 10^24 kg. Every
value has to be in that one unit: entered as 1.898 (in 10^27 kg), Jupiter would come out 1000
times too light in every comparison and ratio below.

In [51]:
planet_masses_1e24kg= {
    "mercury": 0.330,
    "venus": 4.87,
    "earth": 5.97,
    "mars":0.642,
    "jupiter":1898,
    "saturn":568,
    "uranus":86.8,
    "neptune":102
}

**Q6) Use your dictionary to look up Earth's mass.**

In [52]:
planet_masses_1e24kg.get("earth")

5.97

**Q7) Loop through the dictionary and build a list of every planet heavier than
100 x 10^24 kg.**

You should end up with `["jupiter", "saturn", "neptune"]` — uranus, at 86.8, does not make it.

Hint: `.items()` gives you the name and the mass together on each pass, as in 1.2's lecture.

In [54]:
heavier_planets= []
for planet, mass  in planet_masses_1e24kg.items():
  if mass > 100:
    heavier_planets.append(planet)
print(heavier_planets)

['jupiter', 'saturn', 'neptune']


**Q8) Add Pluto to your dictionary.**

Its mass is 0.0130 x 10^24 kg. Print the names afterwards, looping with `.items()`, to confirm it
landed.

Pluto has not been a planet since 2006. Adding it still takes one assignment, and none of the code
that reads the other eight entries has to change.

In [57]:
planet_masses_1e24kg["pluto"] = 0.0130

for planet, mass in planet_masses_1e24kg.items():
    print(planet)

mercury
venus
earth
mars
jupiter
saturn
uranus
neptune
pluto


### C: functions

**Q9) Write a function that converts a planetary mass to Earth masses.**

Jupiter is about 1898 x 10^24 kg, which is about 318 Earth masses:

$$M_{\text{Jupiter}} \approx 1898 \times 10^{24}\,\text{kg} \approx 318\,M_{\text{Earth}}$$

Call it `to_earth_masses(mass_1e24kg)` and check it against Jupiter — you should get
317.92294807370183.

In [58]:
def to_earth_masses(mass_1e24kg):
    return mass_1e24kg / planet_masses_1e24kg["earth"]

jupiter_earth_masses = to_earth_masses(planet_masses_1e24kg["jupiter"])

print(jupiter_earth_masses)


317.92294807370183


**Q10) Now write a single function that converts to either Earth masses or Jupiter
masses, depending on a keyword argument.**

Call it `to_planet_masses(mass_1e24kg, reference="jupiter")`. Converting Jupiter's own mass should
give 317.923 with `reference="earth"` and exactly 1.0 with `reference="jupiter"`.

With the default argument, one function covers both conversions. Jupiter in Jupiter masses has
to come out as exactly 1.0; any other value means the function divides by the wrong mass.

In [59]:
def to_planet_masses(mass_1e24kg, reference="jupiter"):
    return mass_1e24kg / planet_masses_1e24kg[reference]


jupiter_earth = to_planet_masses(
    planet_masses_1e24kg["jupiter"],
    reference="earth"
)

jupiter_jupiter = to_planet_masses(
    planet_masses_1e24kg["jupiter"],
    reference="jupiter"
)

print(jupiter_earth)
print(jupiter_jupiter)

317.92294807370183
1.0


**Q11) Write a function that returns two values.**

`mass_in_two_units(mass_1e24kg, reference)` should return the mass in Earth masses *and* in masses
of whatever reference planet you pass. Test it on Jupiter with `reference="mars"`.

Hint: `return a, b` returns a tuple, and `first, second = f(...)` unpacks it — both from 1.2's
lecture.

In [61]:
def mass_in_two_units(mass_1e24kg, reference):
    earth_masses = mass_1e24kg / planet_masses_1e24kg["earth"]
    reference_masses = mass_1e24kg / planet_masses_1e24kg[reference]

    return earth_masses, reference_masses
earth_masses, mars_masses = mass_in_two_units(
    planet_masses_1e24kg["jupiter"],
    reference="mars"
)

print(earth_masses)
print(mars_masses)

317.92294807370183
2956.386292834891


**Bonus.** Convert Neptune's mass to Jupiter masses with your Q10 function. Then pass that result
to your Q11 function with `reference="jupiter"`, and compare the second value it returns with the
Q10 result. Are they the same? If not, work out which function is being given a number in the
wrong unit, and what it should have been given instead — this is the mistake the `_1e24kg` suffix
exists to prevent.

In [62]:
neptune_jupiter = to_planet_masses(
    planet_masses_1e24kg["neptune"],
    reference="jupiter"
)

print(neptune_jupiter)



earth_masses, jupiter_masses = mass_in_two_units(
    neptune_jupiter,
    reference="jupiter"
)

print(jupiter_masses)

0.05374077976817703
2.831442558913437e-05


**ℹ️ What you just did**

You built a list and read its length rather than counting it, sliced a subset out of it, and
looped over it with a condition. You built a dictionary keyed by name, looked a value up, filtered
it into a new list, and extended it. Then you wrote three functions over that dictionary: one
fixed, one with a default argument that collapses two functions into one, and one returning a pair.

That is the whole of this subchapter applied to one dataset. In 1.7 you will come back and replace
the dictionary with a class, so that a planet's name and mass travel together as one object
instead of as a key and a value that happen to line up.